File: mpi/S2/t1/ex2a.f
      implicit REAL*8 (A-H,O-Z)
      include  'mpif.h'

      integer(kind=4) :: my_rank, PETOT, NEIB
      real   (kind=8) :: VEC(36)

      integer(kind=4), dimension(MPI_STATUS_SIZE,1) :: stat_send
      integer(kind=4), dimension(MPI_STATUS_SIZE,1) :: stat_recv
      integer(kind=4), dimension(1)    :: request_send
      integer(kind=4), dimension(1)    :: request_recv

      integer(kind=4) :: start_send, length_send
      integer(kind=4) :: start_recv, length_recv

      call MPI_INIT      (ierr)
      call MPI_COMM_SIZE (MPI_COMM_WORLD, PETOT, ierr )
      call MPI_COMM_RANK (MPI_COMM_WORLD, my_rank, ierr )

      if (my_rank.eq.0) then
        NEIB= 1
        start_send=   1
        length_send=  11
        start_recv=  length_send + 1
        length_recv=  25
        do i= 1, 36
          VEC(i)= 100 + i
        enddo
      endif

      if (my_rank.eq.1) then
        NEIB= 0
         start_send=  1
        length_send= 25
         start_recv= length_send + 1
        length_recv= 11
        do i= 1, 36
          VEC(i)= 200 + i
        enddo
      endif

      do i= 1, 36
        write (*,'(i1,a,i2,f10.0)') my_rank, ' #BEFORE# ', i,VEC(i)
      enddo

      call MPI_ISEND (VEC(start_send), length_send,                     &
     &                MPI_DOUBLE_PRECISION, NEIB, 0, MPI_COMM_WORLD,    &
     &                request_send(1), ierr)
      call MPI_IRECV (VEC(start_recv), length_recv,                     &
     &                MPI_DOUBLE_PRECISION, NEIB, 0, MPI_COMM_WORLD,    &
     &                request_recv(1), ierr)

      call MPI_WAITALL (1, request_recv, stat_recv, ierr)
      call MPI_WAITALL (1, request_send, stat_send, ierr)
   
      do i= 1, 36
        write (*,'(i1,a,i2,f10.0)') my_rank, ' #AFTER # ', i,VEC(i)
      enddo

      call MPI_FINALIZE (ierr)

      end 

File: mpi/S2/t1/ex2a.c
#include <stdio.h>
#include <stdlib.h>
#include "mpi.h"
int main(int argc, char **argv){
	int i, neib;
	int MyRank, PeTot;
	double VEC[36];
	int Start_Send, Length_Send;
	int Start_Recv, Length_Recv;

	MPI_Status *StatSend, *StatRecv;
	MPI_Request *RequestSend, *RequestRecv;

	MPI_Init(&argc, &argv);
	MPI_Comm_size(MPI_COMM_WORLD, &PeTot);
	MPI_Comm_rank(MPI_COMM_WORLD, &MyRank);

        Start_Send= 1;

	if(MyRank == 0) {
	  neib= 1;      
          Length_Send= 11;
          Length_Recv= 25;  
          for (i=0;i<36;i++){VEC[i]=100.0+i+1;}}

          if(MyRank == 1) {
	  neib= 0;      
          Length_Send= 25;
          Length_Recv= 11;  
          for (i=0;i<36;i++){VEC[i]=200.0+i+1;}}

        Start_Recv= 1 + Length_Send;

	StatSend = malloc(sizeof(MPI_Status) * 1);
	StatRecv = malloc(sizeof(MPI_Status) * 1);
	RequestSend = malloc(sizeof(MPI_Request) * 1);
	RequestRecv = malloc(sizeof(MPI_Request) * 1);

        for (i=0;i<36;i++) {
	  printf("%s%2d%5d%8.0f\n", "### before", MyRank, i+1, VEC[i]);}

	MPI_Isend(&VEC[Start_Send-1], Length_Send, MPI_DOUBLE, neib, 0, 
                   MPI_COMM_WORLD, &RequestSend[0]);
	MPI_Irecv(&VEC[Start_Recv-1], Length_Recv, MPI_DOUBLE, neib, 0, 
                   MPI_COMM_WORLD, &RequestRecv[0]);
        MPI_Waitall(1, RequestRecv, StatRecv);
        MPI_Waitall(1, RequestSend, StatSend);

        for (i=0;i<36;i++) {
	  printf("%s%2d%5d%8.0f\n", "### after ", MyRank, i+1, VEC[i]);}
	
	MPI_Finalize();
	return 0;
}

